releng/release-tool/release-tool.go

```diff
@@ -0,0 +1,1186 @@
+			// -----------------------------------------------------------------------------------
+			if !blocksNewRC {
+				highestRC++
+				nextMinorReleaseRC = fmt.Sprintf("%s%d", rcTemplate, highestRC)
+				shouldMakeNewRC = true
+			}
+		} else if secondsDiff >= promoteAfterSeconds {
+
+			// ------------------------------------------------------------------------------
+			// 7. Detect if it's time to promote a x.y.0.rc.n release to an official release.
+			// ------------------------------------------------------------------------------
+
+			shouldPromoteRC = true
+		} else {
+			log.Printf("Waiting to promote RC %s. %d seconds remain", *rcPromotionCandidate.TagName, promoteAfterSeconds-secondsDiff)
+		}
+	}
+
+	if shouldMakeNewMinorRelease {
+		log.Printf("Auto mode detected a new branch [%s] and new tag [%s] should be created", nextMinorReleaseBranch, nextMinorReleaseRC)
+
+		r.tag = nextMinorReleaseRC
+		r.newBranch = nextMinorReleaseBranch
+		r.tagBranch = nextMinorReleaseBranch
+
+		return nil
+	} else if shouldMakeNewRC {
+		log.Printf("Auto creating new RC [%s] for branch [%s]", nextMinorReleaseRC, nextMinorReleaseBranch)
+
+		r.tag = nextMinorReleaseRC
+		r.tagBranch = nextMinorReleaseBranch
+
+		return nil
+	} else if shouldPromoteRC {
+		log.Printf("Auto promoting rc [%s] after [%d] days", *rcPromotionCandidate.TagName, autoPromoteAfterDays)
+		r.promoteRC = *rcPromotionCandidate.TagName
+		return nil
+	}
+
+	log.Printf("No auto action detected for %s/%s", r.org, r.repo)
+	return nil
+}
+
+func (r *releaseData) verifyTag() error {
+	// must be a valid semver version
+	tagSemver, err := semver.NewVersion(r.tag)
+	if err != nil {
+		return err
+	}
+
+	expectedBranch := fmt.Sprintf("release-%d.%d", tagSemver.Major(), tagSemver.Minor())
+
+	releases, err := r.getReleases()
+	for _, release := range releases {
+		if *release.TagName == r.tag {
+			log.Printf("Release tag [%s] already exists", r.tag)
+			return nil
+		}
+	}
+
+	// ensure promoteRC tag exists and is eligible for promotion
+	if r.promoteRC != "" {
+		found := false
+		invalid := false
+		for _, release := range releases {
+			if *release.TagName == r.promoteRC {
+				found = true
+				invalid, _, err = r.isRCInvalid(release, expectedBranch)
+				if err != nil {
+					return err
+				}
+				r.promoteRCTime = release.CreatedAt.Time
+				break
+			}
+		}
+		if !found {
+			return fmt.Errorf("Unable to find promote-rc tag [%s]", r.promoteRC)
+		} else if invalid {
+			return fmt.Errorf("RC [%s] is ineligible to be promoted due to blockers.", r.promoteRC)
+		}
+	} else {
+		// if this is not a promotion, ensure the release is either an RC or a patch release.
+		re := regexp.MustCompile(`^v\d*\.\d*.\d*-rc.\d*$`)
+		match := re.FindString(r.tag)
+		if match == "" && tagSemver.Patch() == 0 && !r.force {
+			return fmt.Errorf("The tag [%s] must be promoted from a release candidate since it is the first release of a patch series.", r.tag)
```

Comment:
Maybe it makes sense to have a defined exit codes in those cases? it will allow us to run this tool automatically and handle different errors.

Reply:
the automation parts are meant to handle all this already. If we hit an error like this, it would be because a user tried to do something unexpected.